src/main/java/net/minecraftforge/client/model/ForgeBlockStateV1.java

```diff
@@ -0,0 +1,519 @@
+
+            if (vars != null)
+            {
+                for (ForgeBlockStateV1.Variant v : vars)
+                {
+                    if (v != null)
+                    {   // We put this part variant in the permutation's map to add further in recursion, and then remove it afterward just in case.
+                        parts.put(name, v);
+                        getSubmodelPermutations(baseVar, sorted, map, depth + 1, parts, ret);
+                        parts.remove(name);
+                    }
+                }
+            }
+            else
+            {
+                getSubmodelPermutations(baseVar, sorted, map, depth + 1, parts, ret);
+            }
+
+            return ret;
+        }
+
+        private List<ForgeBlockStateV1.Variant> getSubmodelPermutations(ForgeBlockStateV1.Variant baseVar, Map<String, List<ForgeBlockStateV1.Variant>> variants)
+        {
+            List<String> sorted = Lists.newArrayList(variants.keySet());
+            Collections.sort(sorted);   // Sort to get consistent results.
+            return getSubmodelPermutations(baseVar, sorted, variants, 0, new HashMap<String, ForgeBlockStateV1.Variant>(), new ArrayList<ForgeBlockStateV1.Variant>());
+        }
+    }
+
+    public static class Variant
+    {
+        public static final Object SET_VALUE = new Object();
+
+        private ResourceLocation model = null;
+        private boolean modelSet = false;
+        private Optional<ModelRotation> rotation = Optional.absent();
+        private Optional<Boolean> uvLock = Optional.absent();
+        private Optional<Integer> weight = Optional.absent();
+        private Map<String, String> textures = Maps.newHashMap();
+        private Map<String, List<ForgeBlockStateV1.Variant>> submodels = Maps.newHashMap();
+        private Map<String, Object> simpleSubmodels = Maps.newHashMap(); // Makeshift Set to allow us to "remove" (replace value with null) singleParts when needed.
+        private Map<String, String> customData = Maps.newHashMap();
+
+        private Variant(){}
+        /**
+         * Clone a variant.
+         * @param other Variant to clone.
+         */
+        private Variant(ForgeBlockStateV1.Variant other)
+        {
+            this.model = other.model;
+            this.modelSet = other.modelSet;
+            this.rotation = other.rotation;
+            this.uvLock = other.uvLock;
+            this.weight = other.weight;
+            this.textures.putAll(other.textures);
+            this.submodels.putAll(other.submodels);
+            this.simpleSubmodels.putAll(other.simpleSubmodels);
+            this.customData.putAll(other.customData);
+        }
+
+        /**
+         * Sets values in this variant to the input's values only if they haven't been set already. Essentially inherits values from o.
+         */
+        ForgeBlockStateV1.Variant sync(ForgeBlockStateV1.Variant parent)
+        {
+            if (!this.modelSet)             this.model    = parent.model;
```

Comment:
this.modelSet is never changed, which causes improper model overwrites.
Adding the line bellow after the model is set to the parent's model corrects this.
if (!this.modelSet) this.modelSet = parent.modelSet;